core/src/autogluon/core/utils/feature_selection.py

```diff
@@ -0,0 +1,497 @@
+            candidate_found = candidate_set != best_set and candidate_set != prev_candidate_set and len(candidate_set) > 0
+            all_features_evaluated = len(importance_df[importance_df['importance'].isna()]) == 0
+            fn_args['prev_importance_df'] = importance_df
+            if candidate_found or all_features_evaluated:
+                break
+        logger.log(30, f"\tCandidate generation time: ({round(total_prune_time, 2)}s), Cardinality: {len(candidate_features)}")
+        return candidate_features, importance_df, candidate_found, total_prune_time
+
+    def compute_next_candidate_round(self, X: pd.DataFrame, y: pd.Series, model: AbstractModel, time_budget: float, features: List[str],
+                                     n_subsample: int, min_fi_samples: int, max_fi_samples: int, prune_ratio: float, prune_threshold: float,
+                                     prev_importance_df: pd.DataFrame = None, prioritized: Set[str] = set(), using_prev_fit_fi: Set[str] = set(),
+                                     weighted: bool = True) -> Tuple[List[str], pd.DataFrame, float]:
+        """
+        Compute permutation feature importance for as many features as possible under time_budget. Ensure each returned feature importance
+        scores are computed from at least n_sample datapoints. Update self.time_limit to account for feature importance computation time.
+        """
+        # determine how many subsamples and shuffles to use for feature importance calculation
+        time_start = time.time()
+        n_features = len(features)
+        n_subsample = min(n_subsample, len(X))
+        n_sample = max(min_fi_samples, min(max_fi_samples, len(X)))
+        n_shuffle = min(np.ceil(n_sample / n_subsample).astype(int), self.max_n_shuffle)
+        expected_single_feature_time = self.compute_expected_fi_time_single(X, self.model_predict_time, n_subsample, n_sample)
+        auto_threshold = len(prioritized) > 0
+        features = sort_features_by_priority(features, prioritized, prev_importance_df, using_prev_fit_fi)
+
+        # if we do not have enough time to evaluate feature importance for all features, do so only for some (first n_evaluated_features elements of features)
+        n_evaluated_features = max([i for i in range(0, n_features+1) if i * expected_single_feature_time <= time_budget])
+        if n_evaluated_features == 0:
+            prune_time = time.time() - time_start
+            self.time_limit = self.time_limit - prune_time
+            self._fi_time_elapsed = self._fi_time_elapsed + prune_time
+            return features, unevaluated_fi_df_template(features), prune_time
+        evaluated_features = features[:n_evaluated_features]
+        unevaluated_features = features[n_evaluated_features:]
+        time_budget_fi = time_budget - (time.time() - time_start)
+        logger.log(30, f"\tComputing feature importance for {n_evaluated_features}/{n_features} features with {n_shuffle} shuffles.")
```

Comment:
FYI logging level 30 is supposed to be reserved for important statements. Regular statements should be logging level 20, and statements only related to internal operations should probably be left at 15 so they only appear for users who've purposefully set the verbosity level to be higher (to avoid overwhelming them)